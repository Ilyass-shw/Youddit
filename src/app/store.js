import { configureStore } from '@reduxjs/toolkit';
import postsReducer from '../features/posts/postsSlice/postsSlice';

export default configureStore({
  reducer: {
    posts: postsReducer,
  },
});
