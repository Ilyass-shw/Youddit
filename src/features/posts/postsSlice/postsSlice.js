import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import { client, getEndPoint, handlefetchedPosts } from "../../../api/api";

const initialState = {
	searchTerm: "memes",
	postList: [],
	filter: "relevance",
	firstFilter: "relevance",
	lastFilter: "comments",
	postFetchingStatus: "idle",
	error: null,
};

export const fetchPosts = createAsyncThunk("posts/fetchPosts", async (_, { getState }) => {
	const filter = getState().posts.filter;
	const searchTerm = getState().posts.searchTerm;

	const endPoint = getEndPoint(searchTerm, filter);
	const response = await client.get(endPoint);

	const fetchedPosts = handlefetchedPosts(response);
	return fetchedPosts;
});

const postsSlice = createSlice({
	name: "posts",
	initialState,
	reducers: {
		searchTermSet(state, action) {
			state.searchTerm = action.payload;
		},
		filterUpdated(state, action) {
			state.filter = action.payload;
		},
		firstFilterUpdated(state, action) {
			state.firstFilter = action.payload;
		},
		lastFilterUpdated(state, action) {
			state.lastFilter = action.payload;
		},
	},
	extraReducers: {
		[fetchPosts.pending]: (state, action) => {
			state.postFetchingStatus = "Loading";
		},
		[fetchPosts.fulfilled]: (state, action) => {
			state.postFetchingStatus = "succeeded";
			state.postList = action.payload;
		},
		[fetchPosts.rejected]: (state, action) => {
			state.postFetchingStatus = "failed";
			state.error = action.error.message;
		},
	},
});

export const { searchTermSet, filterUpdated, firstFilterUpdated, lastFilterUpdated } = postsSlice.actions;
export default postsSlice.reducer;

export const selectAllPosts = (state) => state.posts.postList;
