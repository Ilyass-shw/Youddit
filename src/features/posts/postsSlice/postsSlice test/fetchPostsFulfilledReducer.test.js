import {  fetchPosts } from "../postsSlice";
import store from "../../../../app/store.js";
import { wait } from "@testing-library/react";

describe("fetchPosts fulfilled reducer", () => {
	it("should change postFetchingStatus from idle to succeeded if the fetching succeeds", async () => {
		let postFetchingStatus = store.getState().posts.postFetchingStatus;
		let postList = store.getState().posts.postList;

		expect(postFetchingStatus).toBe("idle");
		expect(postList).toHaveLength(0);

		store.dispatch(fetchPosts());

		await wait(() => expect(store.getState().posts.postFetchingStatus).toBe("succeeded"));

		postList = store.getState().posts.postList;

		expect(postList).toHaveLength(3);
		expect(postList).toMatchSnapshot();
	});
});
