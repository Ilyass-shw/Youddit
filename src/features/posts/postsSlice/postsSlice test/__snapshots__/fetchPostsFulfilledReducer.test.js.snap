// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchPosts fulfilled reducer should change postFetchingStatus from idle to succeeded if the fetching succeeds 1`] = `
Array [
  Object {
    "author": "author 1",
    "date": "16 days ago",
    "id": "id 1",
    "img": "fake img url 1",
    "subreddit": "subreddit 1",
    "title": "fake title 1",
    "upvotes": "115.6K",
    "url": "fake link 1",
  },
  Object {
    "author": "author 2",
    "date": "in over 63 years",
    "id": "id 2",
    "img": "fake img url 2",
    "subreddit": "subreddit 2",
    "title": "fake title 2",
    "upvotes": "115.51K",
    "url": "fake link 2",
  },
  Object {
    "author": "author 3",
    "date": "in over 31 years",
    "id": "id 3",
    "img": "fake img url 3",
    "subreddit": "subreddit 3",
    "title": "fake title 3",
    "upvotes": "115.59K",
    "url": "fake link 3",
  },
]
`;
